Worker stop request is lost (deadlock) when cancellation lands during startup

## Symptom

`test_stop_worker_aborts_sync_jobs_past_shutdown_graceful_timeout` failed on main in CI ([run](https://github.com/procrastinate-org/procrastinate/actions/runs/27297387907/job/80633639248)):

```
assert fast_job_status == Status.SUCCEEDED
E       AssertionError: assert <Status.TODO: 'todo'> == <Status.SUCCEEDED: 'succeeded'>
```

Investigating the flake surfaced a more serious underlying bug: a stop request (or task cancellation) landing in the worker's startup window is silently erased, leaving the worker unstoppable — and, with a sync job waiting on `should_abort()`, deadlocked.

## Bug 1: `_run_loop`'s `_stop_event.clear()` erases an early stop request

`Worker._run_loop` starts with `self._stop_event.clear()`. The fatal interleaving:

1. `run()` creates `loop_task = asyncio.create_task(self._run_loop())` — scheduled, not yet started
2. Cancellation hits `run()` at `await asyncio.shield(loop_task)`; the `except CancelledError` handler calls `self.stop()`, which sets `_stop_event`
3. `run()` proceeds to `await loop_task`; only now does `_run_loop` get its first slot — and its first statement clears the just-set event
4. The worker runs as if no stop was requested: graceful shutdown (and the abort it would issue) never begins. A sync job polling `context.should_abort()` spins forever, and `await run_task` never returns

Reproduced by shrinking the pre-cancel `asyncio.sleep(0.05)` in the test above to `0.001`: the test hangs instead of failing. py-spy of the hung process shows the event loop idle and `slow_job` spinning in its worker thread waiting for an abort that never comes.

## Bug 2 (the flake actually seen in CI): fixed 50 ms sleep races worker startup

The `*_past_shutdown_graceful_timeout` tests sleep a fixed `0.05s` between starting the worker and cancelling it. In that time the worker must prune stalled workers, register itself, start the run loop and side tasks, fetch and run the fast job, and fetch the slow job — several DB roundtrips. On a loaded CI runner the cancel wins the race: graceful shutdown finds no running job and exits cleanly, the jobs were never fetched, hence `TODO`. (When the cancel lands slightly later, you get the bug-1 hang instead.)

## Fix

Branch `fix-worker-stop-race` (PR to follow):

- Move the `_stop_event.clear()` out of `_run_loop` to the top of `run()`, before the first await, so no stop requested after `run()` has started can be erased. Behavior is otherwise unchanged (a stop requested before `run()` is still discarded, worker instances stay reusable).
- Deterministic regression test for the lost-stop window.
- De-flake the four shutdown acceptance tests by polling for the slow job reaching `doing` (new `wait_for_job_status` helper) instead of sleeping a fixed 50 ms.

The `ValueError: I/O operation on closed file` logging tracebacks in the CI output are pytest capture noise around the failure, not a procrastinate bug.

--- a/procrastinate/worker.py
+++ b/procrastinate/worker.py
@@ -421,6 +421,11 @@ class Worker:
         Run the worker
         This will run forever until asked to stop/cancelled, or until no more job is available is configured not to wait
         """
+        # Reset a stale stop request from a previous run. This must happen before
+        # the first await: anything later (e.g. the top of the run loop task) can
+        # run after a fresh stop() and erase it, leaving the worker unstoppable.
+        self._stop_event.clear()
+
         logger.debug("Pruning stalled workers with old heartbeats")
         pruned_workers = await self.app.job_manager.prune_stalled_workers(
             self.stalled_worker_timeout
@@ -640,7 +645,6 @@ class Worker:
             ),
         )
         self._new_job_event.clear()
-        self._stop_event.clear()
         self._running_jobs = {}
         self._job_semaphore = asyncio.Semaphore(self.concurrency)
         side_tasks = self._start_side_tasks()

--- a/tests/acceptance/test_async.py
+++ b/tests/acceptance/test_async.py
@@ -24,6 +24,16 @@ async def async_app(request, psycopg_connector, connection_params):
         yield app
 
 
+async def wait_for_job_status(
+    app: app_module.App, job_id: int, status: Status, timeout: float = 5
+):
+    async def poll():
+        while await app.job_manager.get_job_status_async(job_id) != status:
+            await asyncio.sleep(0.02)
+
+    await asyncio.wait_for(poll(), timeout)
+
+
 async def test_defer(async_app: app_module.App):
     sum_results = []
     product_results = []
@@ -349,7 +359,7 @@ async def test_stop_worker_aborts_async_jobs_past_shutdown_graceful_timeout(
     run_task = asyncio.create_task(
         async_app.run_worker_async(wait=False, shutdown_graceful_timeout=0.3)
     )
-    await asyncio.sleep(0.05)
+    await wait_for_job_status(async_app, slow_job_id, Status.DOING)
 
     with pytest.raises(asyncio.CancelledError):
         run_task.cancel()
@@ -382,7 +392,7 @@ async def test_stop_worker_retries_async_jobs_past_shutdown_graceful_timeout(
     run_task = asyncio.create_task(
         async_app.run_worker_async(wait=False, shutdown_graceful_timeout=0.3)
     )
-    await asyncio.sleep(0.05)
+    await wait_for_job_status(async_app, slow_job_id, Status.DOING)
 
     with pytest.raises(asyncio.CancelledError):
         run_task.cancel()
@@ -417,7 +427,7 @@ async def test_stop_worker_aborts_sync_jobs_past_shutdown_graceful_timeout(
     run_task = asyncio.create_task(
         async_app.run_worker_async(wait=False, shutdown_graceful_timeout=0.3)
     )
-    await asyncio.sleep(0.05)
+    await wait_for_job_status(async_app, slow_job_id, Status.DOING)
 
     with pytest.raises(asyncio.CancelledError):
         run_task.cancel()
@@ -452,7 +462,7 @@ async def test_stop_worker_retries_sync_jobs_past_shutdown_graceful_timeout(
     run_task = asyncio.create_task(
         async_app.run_worker_async(wait=False, shutdown_graceful_timeout=0.3)
     )
-    await asyncio.sleep(0.05)
+    await wait_for_job_status(async_app, slow_job_id, Status.DOING)
 
     with pytest.raises(asyncio.CancelledError):
         run_task.cancel()

--- a/tests/acceptance/test_async.py
+++ b/tests/acceptance/test_async.py
@@ -149,11 +149,15 @@ async def test_no_job_to_cancel_found(async_app: app_module.App):
 async def test_abort_async_task(async_app: app_module.App, mode):
     @async_app.task(queue="default", name="task1")
     async def task1():
-        await asyncio.sleep(0.5)
+        # Outlasts the wait_for timeouts below, so the abort always wins; on
+        # success the sleep is cancelled and the duration never elapses.
+        await asyncio.sleep(3)
 
     job_id = await task1.defer_async()
 
-    abort_job_polling_interval = 0.1
+    # In listen mode, the polling interval is far beyond the wait_for timeout
+    # below, so the test can only pass through the notification path.
+    abort_job_polling_interval = 0.1 if mode == "poll" else 30
 
     worker_task = asyncio.create_task(
         async_app.run_worker_async(
@@ -164,15 +168,13 @@ async def test_abort_async_task(async_app: app_module.App, mode):
         )
     )
 
-    await asyncio.sleep(0.05)
+    await wait_for_job_status(async_app, job_id, Status.DOING)
     result = await async_app.job_manager.cancel_job_by_id_async(job_id, abort=True)
     assert result is True
 
-    # when listening for notifications, job should cancel within ms
-    # if notifications are disabled, job will only cancel after abort_job_polling_interval
-    await asyncio.wait_for(
-        worker_task, timeout=0.1 if mode == "listen" else abort_job_polling_interval * 2
-    )
+    # Generous multiples of the expected latency (notification: ms, polling: one
+    # 0.1s interval) so a slow CI runner doesn't fail the test.
+    await asyncio.wait_for(worker_task, timeout=1 if mode == "listen" else 2)
 
     status = await async_app.job_manager.get_job_status_async(job_id)
     assert status == Status.ABORTED
@@ -189,7 +191,9 @@ async def test_abort_sync_task(async_app: app_module.App, mode):
 
     job_id = await task1.defer_async()
 
-    abort_job_polling_interval = 0.1
+    # In listen mode, the polling interval is far beyond the wait_for timeout
+    # below, so the test can only pass through the notification path.
+    abort_job_polling_interval = 0.1 if mode == "poll" else 30
 
     worker_task = asyncio.create_task(
         async_app.run_worker_async(
@@ -200,15 +204,13 @@ async def test_abort_sync_task(async_app: app_module.App, mode):
         )
     )
 
-    await asyncio.sleep(0.05)
+    await wait_for_job_status(async_app, job_id, Status.DOING)
     result = await async_app.job_manager.cancel_job_by_id_async(job_id, abort=True)
     assert result is True
 
-    # when listening for notifications, job should cancel within ms
-    # if notifications are disabled, job will only cancel after abort_job_polling_interval
-    await asyncio.wait_for(
-        worker_task, timeout=0.1 if mode == "listen" else abort_job_polling_interval * 2
-    )
+    # Generous multiples of the expected latency (notification: ms, polling: one
+    # 0.1s interval) so a slow CI runner doesn't fail the test.
+    await asyncio.wait_for(worker_task, timeout=1 if mode == "listen" else 2)
 
     status = await async_app.job_manager.get_job_status_async(job_id)
     assert status == Status.ABORTED

--- a/tests/acceptance/test_async.py
+++ b/tests/acceptance/test_async.py
@@ -326,7 +326,8 @@ async def test_stop_worker(async_app: app_module.App):
     job_ids.append(await appender.defer_async(a=2))
 
     run_task = asyncio.create_task(async_app.run_worker_async(concurrency=2, wait=True))
-    await asyncio.sleep(0.5)
+    for job_id in job_ids:
+        await wait_for_job_status(async_app, job_id, Status.SUCCEEDED)
 
     with pytest.raises(asyncio.CancelledError):
         run_task.cancel()
@@ -434,8 +435,6 @@ async def test_stop_worker_aborts_sync_jobs_past_shutdown_graceful_timeout(
     with pytest.raises(asyncio.CancelledError):
         run_task.cancel()
         await run_task
-
-    await asyncio.sleep(0.05)
 
     fast_job_status = await async_app.job_manager.get_job_status_async(fast_job_id)
     slow_job_status = await async_app.job_manager.get_job_status_async(slow_job_id)

--- a/tests/unit/test_worker.py
+++ b/tests/unit/test_worker.py
@@ -84,6 +84,23 @@ async def test_worker_run_wait_stop(app: App, caplog):
     }
 
 
+async def test_stop_requested_before_run_loop_starts_is_not_lost(app: App):
+    worker = Worker(app, wait=True, install_signal_handlers=False)
+
+    original_run_loop = worker._run_loop
+
+    async def run_loop_with_late_start():
+        # Simulate a stop() landing after run() scheduled the run loop task but
+        # before the task's first statement; if it gets erased, with wait=True
+        # the worker runs forever.
+        worker.stop()
+        await original_run_loop()
+
+    worker._run_loop = run_loop_with_late_start
+
+    await asyncio.wait_for(worker.run(), timeout=2)
+
+
 def test_stop_after_run_exited_with_error_does_not_raise(not_opened_app: App):
     # If run() dies on an exception (e.g. a database error), the stop event is
     # never set and the worker's event loop ends up closed. A late stop() (e.g.

